packages/babel-core/src/errors/rewrite-stack-trace.ts

```diff
@@ -0,0 +1,167 @@
+ * - If c() throws an unexpected error, then its shown call stack will be "c, d, e, f"
+ * - If d() throws an expected error, then its shown call stack will be "e, f"
+ * - If d() throws an unexpected error, then its shown call stack will be "d, e, f"
+ * - If e() throws an error, then its shown call stack will be "e, f"
+ *
+ * Additionally, an error can inject additional "virtual" stack frames using the
+ * injcectVirtualStackFrame(error, filename) function: those are added on the top
+ * of the existig stack, after hiding the possibly hidden frames.
+ * In the example above, if we called injcectVirtualStackFrame(error, "h") on the
+ * expected error thrown by c(), it's shown call stack would have been "h, e, f".
+ * This can be useful, for example, to report config validation errors as if they
+ * were directly thrown in the config file.
+ */
+
+const ErrorToString = Function.call.bind(Error.prototype.toString);
+
+const SUPPORTED = !!Error.captureStackTrace;
+
+// We add some extra frames to Error.stackTraceLimit, so that we can respect
+// the original Error.stackTraceLimit even after removing all our internal
+// frames.
+// STACK_TRACE_LIMIT_DELTA should be bigger than the expected number of internal
+// frames, but not too big because capturing the stack trace is slow (this is
+// why Error.stackTraceLimit does not default to Infinity!).
+// Increase it if needed.
+const STACK_TRACE_LIMIT_DELTA = 100;
+
+const START_HIDNG = "startHiding - secret - don't use this - v1";
+const STOP_HIDNG = "stopHiding - secret - don't use this - v1";
+
+type CallSite = Parameters<typeof Error.prepareStackTrace>[1][number];
+
+const expectedErrors = new WeakSet<Error>();
+const virtualFrames = new WeakMap<Error, CallSite[]>();
+
+function CallSite(filename: string): CallSite {
+  // We need to use a prototype otherwise it breaks source-map-support's internals
+  return Object.create({
+    isNative: () => false,
+    isConstructor: () => false,
+    isToplevel: () => true,
+    getFileName: () => filename,
+    getLineNumber: () => undefined,
+    getColumnNumber: () => undefined,
+    getFunctionName: () => undefined,
+    getMethodName: () => undefined,
+    getTypeName: () => undefined,
+    toString: () => filename,
+  } as CallSite);
+}
+
+export function injcectVirtualStackFrame(error: Error, filename: string) {
+  if (!SUPPORTED) return;
+
+  let frames = virtualFrames.get(error);
+  if (!frames) virtualFrames.set(error, (frames = []));
+  frames.push(CallSite(filename));
+
+  return error;
+}
+
+export function expectedError(error: Error) {
+  if (!SUPPORTED) return;
+  expectedErrors.add(error);
+  return error;
+}
+
+export function beginHiddenCallStack<Fn extends Function>(fn: Fn): Fn {
+  if (!SUPPORTED) return fn;
+
+  return Object.defineProperty(
+    function (this: any) {
+      setupPrepareStackTrace();
+      return fn.apply(this, arguments);
+    } as any as Fn,
```

Comment:
The type casting can be avoided if we parameterize the function's input and outputs.

```ts
export function beginHiddenCallStack<T, A extends any[], R>(
  fn: (this: T, ...args: A) => R,
) {
  if (!SUPPORTED) return fn;

  return Object.defineProperty(
    function (this: T, ...args: A) {
      setupPrepareStackTrace();
      return fn.apply(this, args);
    },
    "name",
    { value: STOP_HIDNG },
  );
}
```

Reply:
We still need to type cast somewhere, because the functions that we receive in the config have type `Function` which is not compatible with the "explicit" function types. However, I moved them to the config validation logic so that the "internals" are properly typed.